### PR TITLE
tyr : add API migrate_from_poi_to_osm

### PR DESCRIPTION
**Pull request by mbounabi**
https://jira.canaltp.fr/browse/NAVITIAII-2340
new ticket created afterwards:
https://jira.canaltp.fr/browse/NAVP-756

### CODE REVIEW

File: source/tyr/tyr/api.py [hunk 1]
```diff
@@ -78,6 +78,8 @@
 api.add_resource(resources.AutocompleteUpdateData,
                  '/v0/autocomplete_parameters/<ac_instance_name>/update_data')
 
+api.add_resource(resources.MigrateFromPoiToOsm, '/v0/instances/<string:instance_name>/:actions/migrate_from_poi_to_osm')
```
Comment (pbougue) [inline source/tyr/tyr/api.py]: consistency: In tartare URL is `/actions/<my_action>` https://github.com/CanalTP/tartare/blob/master/documentation/ressources/open_api.yaml#L182

File: source/tyr/tyr/resources.py [hunk 2]
```diff
@@ -1081,3 +1081,18 @@ def post(self, ac_instance_name):
         _, job = import_autocomplete([filename], instance)
         job = models.db.session.merge(job) #reatache the object
         return marshal({'job': job}, one_job_fields), 200
+
+
+class MigrateFromPoiToOsm(flask_restful.Resource):
+    def delete(self, instance_name):
```
Comment (pbougue) [inline source/tyr/tyr/resources.py:1087]: minor consistency: in Tartare, POST method is used https://github.com/CanalTP/tartare/blob/master/documentation/ressources/open_api.yaml#L183
But @kinnou02 is right, it's a delete, so...

File: source/navitiacommon/navitiacommon/models.py [hunk 3]
```diff
@@ -542,15 +548,15 @@ class DataSet(db.Model):
 
     uid = db.Column(UUID, unique=True)
 
-    job_id = db.Column(db.Integer, db.ForeignKey('job.id'))
+    job_id = db.Column(db.Integer, db.ForeignKey('job.id', ondelete='CASCADE'))
```
Comment (kinnou02) [inline source/navitiacommon/navitiacommon/models.py]: This doesn't require a migration?

File: source/tyr/tyr/resources.py [hunk 4]
```diff
@@ -1081,3 +1081,18 @@ def post(self, ac_instance_name):
...
+
+        instance = models.Instance.get_by_name(instance_name)
+        if instance:
+            instance.delete_dataset()
+            return_msg = "All datasets deleted for instance "+instance_name
```
Comment (kinnou02) [inline source/tyr/tyr/resources.py]: Don't concatenate string, always use format like this:
```
"hello {}".format("world")
```
except with logger you should use the logging substitution that use `%s`

File: source/navitiacommon/navitiacommon/models.py [hunk 5]
```diff
@@ -401,6 +401,12 @@ def get_from_id_or_name(cls, id=None, name=None):
         else:
             raise Exception({'error': 'instance is required'}, 400)
 
+    def delete_dataset(self):
+        jobs = db.session.query(Job).filter(Job.instance_id == self.id).all()
+        for j in jobs:
+            db.session.delete(j)
+            db.session.commit()
```
Comment (xlqian) [inline source/navitiacommon/navitiacommon/models.py]:
```python
for j in jobs:
    db.session.delete(j)
db.session.commit()
```

Comment (xlqian) [reply source/navitiacommon/navitiacommon/models.py]: you can probably do this with less code
Job.delete().where(Job.instance_id == self.id)
not sure, to be tested

File: source/tyr/tyr/api.py (hunk 1 above)
Comment (mbounabi) [reply source/tyr/tyr/api.py]: changed to '/v0/instances/<string:instance_name>/actions/migrate_from_poi_to_osm'

File: source/navitiacommon/navitiacommon/models.py (hunk 3 above)
Comment (mbounabi) [reply source/navitiacommon/navitiacommon/models.py]: No upgrade after migration. Now removed.

File: source/tyr/tyr/resources.py (hunk 4 above)
Comment (mbounabi) [reply source/tyr/tyr/resources.py]: Done

File: source/navitiacommon/navitiacommon/models.py (hunk 5 above)
Comment (mbounabi) [reply source/navitiacommon/navitiacommon/models.py]: Done

**Review by TeXitoi — APPROVED** [review]

File: source/navitiacommon/navitiacommon/models.py [hunk 6]
```diff
@@ -401,6 +401,19 @@ def get_from_id_or_name(cls, id=None, name=None):
...
             raise Exception({'error': 'instance is required'}, 400)
 
+    def delete_dataset(self):
+        jobs = db.session.query(Job).filter(Job.instance_id == self.id).all()
+        for j in jobs:
+            for dataset in j.data_sets.all():
+                if dataset.type == 'poi':
+                    db.session.query(Metric).filter(Metric.dataset_id == dataset.id).delete()
+                    db.session.delete(dataset)
+            db.session.commit()
```
Comment (kinnou02) [inline source/navitiacommon/navitiacommon/models.py]: you don't have to commit after every query, you can (and must) commit only at the end

File: source/navitiacommon/navitiacommon/models.py [hunk 7]
```diff
@@ -401,6 +401,19 @@ def get_from_id_or_name(cls, id=None, name=None):
         else:
             raise Exception({'error': 'instance is required'}, 400)
 
+    def delete_dataset(self):
+        jobs = db.session.query(Job).filter(Job.instance_id == self.id).all()
+        for j in jobs:
+            for dataset in j.data_sets.all():
```
Comment (kinnou02) [inline source/navitiacommon/navitiacommon/models.py]: this isn't very efficient, you will get every job then iterate over all the datasets of all these job.
You could do a query to only get the dataset of type POI, then delete

File: source/navitiacommon/navitiacommon/models.py [hunk 8]
```diff
@@ -401,6 +401,19 @@ def get_from_id_or_name(cls, id=None, name=None):
...
+                if dataset.type == 'poi':
+                    db.session.query(Metric).filter(Metric.dataset_id == dataset.id).delete()
+                    db.session.delete(dataset)
+            db.session.commit()
+            # if no more datasets related to job, delete it
+            if not j.data_sets.all():
```
Comment (kinnou02) [inline source/navitiacommon/navitiacommon/models.py]: does it re trigger a query every time?

**Review by kinnou02 — CHANGES_REQUESTED** [review]

File: source/navitiacommon/navitiacommon/models.py (hunk 7 above)
Comment (mbounabi) [reply source/navitiacommon/navitiacommon/models.py]: the new API specifies one instance in the URL. By querying all POI datasets, those from regions not specified in the query will be deleted too

Comment (kinnou02) [reply source/navitiacommon/navitiacommon/models.py]: and you can do `AND` in the query

Comment (mbounabi) [reply source/navitiacommon/navitiacommon/models.py]: Changed

File: source/navitiacommon/navitiacommon/models.py (hunk 6 above)
Comment (mbounabi) [reply source/navitiacommon/navitiacommon/models.py]: Changed

File: source/navitiacommon/navitiacommon/models.py (hunk 8 above)
Comment (mbounabi) [reply source/navitiacommon/navitiacommon/models.py]: according to the doc:
all()
    Return the results represented by this Query as a list.
    This results in an execution of the underlying query.

**Comment by kinnou02** [comment]
BTW it' better to create the branch on your fork, so we don't have a lot of branches on this repo.
it's not bad for this one.

File: source/tyr/tests/integration/migrate_poi_to_osm_test.py [hunk 9]
```diff
@@ -0,0 +1,130 @@
+from tests.check_utils import api_get, api_delete
```
Comment (xlqian) [inline source/tyr/tests/integration/migrate_poi_to_osm_test.py:30]: license

**Review by xlqian — APPROVED** [review]

File: source/navitiacommon/navitiacommon/models.py [hunk 10]
```diff
@@ -401,6 +401,29 @@ def get_from_id_or_name(cls, id=None, name=None):
         else:
             raise Exception({'error': 'instance is required'}, 400)
 
+    def delete_dataset(self):
+        result = db.session.query(DataSet, Job) \
+            .join(Job) \
+            .join(Instance) \
```
Comment (kinnou02) [inline source/navitiacommon/navitiacommon/models.py]: do you really need to do this join? you don't use any fields from instances

Comment (mbounabi) [reply source/navitiacommon/navitiacommon/models.py]: 'self.id' in the filter. Need testing to check the results

File: source/tyr/tests/integration/migrate_poi_to_osm_test.py (hunk 9 above)
Comment (mbounabi) [reply source/tyr/tests/integration/migrate_poi_to_osm_test.py:30]: Done. Only 1 file in the directory contains license header. Should be added in every files if needed

File: source/navitiacommon/navitiacommon/models.py (hunk 10 above)
Comment (mbounabi) [reply source/navitiacommon/navitiacommon/models.py]: Results are the same. Removed

**Review by kinnou02 — APPROVED** [review]

**Pull request merged**